### PR TITLE
Refactor: drop redundant (jruby-complete.jar) dependency

### PR DESCRIPTION
**Pull request by kares**
ATM, LS startup depends on jruby-complete.jar (which includes the Ruby stdlib) from *logstash-core/lib/jars*.
On the other hand, LS has an expanded JRuby installation located at *vendor/jruby*.

## Release notes

[rn:skip]

## What does this PR do?

Changes how LS boots, this boot process is expected to be slightly faster, if noticeable at all.
More importantly it will allow LS to easily patch the standard library (and potentially fix some issues).

## Why is it important/What is the impact to the user?

No impact for the user.

## Author's Checklist

- [x] adjust (and test) Windows .bat scripts 
- [x] ~~restore `custom.jruby.path`~~
- [x] fix license report generation (by providing meta-data for *vendor/jruby/lib/jruby.jar*)

## How to test this PR locally

The usual `bin/logstash ...` way.

## Related issues

- https://github.com/elastic/logstash/issues/12741
- https://github.com/elastic/logstash/issues/12077 (follow-up)

### CODE REVIEW

File: bin/logstash.lib.sh [hunk 1]
```diff
@@ -77,14 +78,18 @@ setup_bundled_jdk_part() {
...
+  local jar_directory="${2}"
   for J in $(cd "${jar_directory}"; ls *.jar); do
     classpath=${classpath}${classpath:+:}${jar_directory}/${J}
   done
   echo "${classpath}"
 }
 
+# set up CLASSPATH as a side-effect, we start to java processes
+CLASSPATH="${JRUBY_HOME}/lib/jruby.jar"
+CLASSPATH="$(setup_classpath $CLASSPATH $LOGSTASH_JARS)"
```
Comment (kares) [inline bin/logstash.lib.sh:91]: we now set up the CP once (as it's used by 2 java invocations)

File: ci/integration_tests.sh [hunk 2]
```diff
@@ -9,7 +9,6 @@ export GRADLE_OPTS="-Xmx4g -Dorg.gradle.jvmargs=-Xmx4g -Dorg.gradle.daemon=false
 
 export SPEC_OPTS="--order rand --format documentation"
 export CI=true
-export OSS=true
```
Comment (kares) [inline ci/integration_tests.sh:12]: unrelated - but this should not longer be needed

File: lib/bootstrap/rspec.rb [hunk 3]
```diff
@@ -20,13 +20,18 @@
...
+  spec_path = File.join(path, "spec")
+  $LOAD_PATH.unshift(spec_path) unless $LOAD_PATH.include?(spec_path)
+end
 
 require "rspec/core"
 require "rspec"
 require 'ci/reporter/rake/rspec_loader'
 
-RSpec.world.reset # if multiple rspec runs occur in a single process, the RSpec "world" state needs to be reset.
+RSpec.clear_examples # if multiple rspec runs occur in a single process, the RSpec "world" state needs to be reset.
```
Comment (kares) [inline lib/bootstrap/rspec.rb:34]: the official way to do a proper RSpec "reset"

File: logstash-core/build.gradle [hunk 4]
```diff
@@ -105,7 +105,7 @@ tasks.register("javaTests", Test) {
 tasks.register("rubyTests", Test) {
     inputs.files fileTree("${projectDir}/lib")
     inputs.files fileTree("${projectDir}/spec")
-    systemProperty 'logstash.core.root.dir', projectDir.absolutePath
+    systemProperty 'logstash.root.dir', projectDir.parent
```
Comment (kares) [inline logstash-core/build.gradle:108]: a new convention for all RSpec Java invocations ...
also `systemProperty 'logstash.core.root.dir'` wasn't previously always pointing to *logstash-core* 
(e.g. from *x-pack/build.gradle*)

File: logstash-core/build.gradle [hunk 5]
```diff
@@ -151,10 +151,9 @@ idea {
     }
 }
 
-def customJRubyDir = project.hasProperty("custom.jruby.path") ? project.property("custom.jruby.path") : ""
```
Comment (kares) [inline logstash-core/build.gradle:154]: *custom.jruby.path* would need more work to get right, 
should be unnecessary as we only need to replace *vendor/jruby* and no longer the complete.jar

File: logstash-core/src/test/java/org/logstash/AccessorsTest.java [hunk 6]
```diff
@@ -32,7 +32,7 @@
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 
-public class AccessorsTest {
+public class AccessorsTest extends RubyTestBase {
```
Comment (kares) [inline logstash-core/src/test/java/org/logstash/AccessorsTest.java:35]: these Java tests caused JRuby runtime initialization ...
-> we would lost control of booting a proper `RubyUtil.RUBY` global runtime

File: qa/integration/framework/fixture.rb [hunk 7]
```diff
@@ -66,16 +66,7 @@ def get_service(name)
...
   end
 
-  def output_equals_expected?
-    FileUtils.identical?(@actual_output, @input)
-  end
-
-  def output_exists?
-    File.exists?(@actual_output)
-  end
-
```
Comment (kares) [inline qa/integration/framework/fixture.rb:76]: defunct - no longer used

File: qa/integration/framework/fixture.rb [hunk 8]
```diff
@@ -66,16 +66,7 @@ def get_service(name)
...
   def teardown
-    File.delete(@actual_output) if @settings.is_set?("actual_output") && output_exists?
```
Comment (kares) [inline qa/integration/framework/fixture.rb:78]: the cleanup was moved to the spec -> as it had assumptions about current work-dir.
(the relative path is considered LS_HOME relative - this is now explicit in the one spec using `actual_output: ...`)

File: logstash-core/src/test/java/org/logstash/config/ir/RubyEnvTestCase.java [hunk 9]
```diff
@@ -44,15 +42,14 @@ public static void before() {
     private static void ensureLoadpath() {
         final LoadService loader = RubyUtil.RUBY.getLoadService();
         if (loader.findFileForLoad("logstash/compiler").library == null) {
+            final String gems = LS_HOME.
```
Comment (yaauie) [inline logstash-core/src/test/java/org/logstash/config/ir/RubyEnvTestCase.java:45]: it seems a variation of our `safePath` helper from elsewhere may come in handy here.

**Review by yaauie — COMMENTED** [review]
First pass: I have made a few comments.

In general, this looks sensible. I'll spend some time ruminating on it and come back later in the week for a full review.

File: logstash-core/src/test/java/org/logstash/common/io/DeadLetterQueueWriterTest.java [hunk 10]
```diff
@@ -200,6 +198,11 @@ public void testCloseFlush() throws Exception {
         }
     }
 
+    private void sleep(long millis) throws InterruptedException {
+        Thread.sleep(millis);
+        Thread.yield();
```
[the remark this answers is not in the document]
Comment (kares) [reply logstash-core/src/test/java/org/logstash/common/io/DeadLetterQueueWriterTest.java:203]: crumbs from my CI fight - was getting failures here, seemed the `sleep` wasn't enough to trigger the other thread.
so I increased the sleep time a bit and added an explicit yield - which has the intent that we're expecting another thread to kick in (subjective preference on mine). if the yield seems confusing I am happy to revert.

File: logstash-core/src/main/java/org/logstash/Logstash.java [hunk 11]
```diff
@@ -215,6 +215,39 @@ public void close() {
...
+    private static RubyInstanceConfig initRubyConfigImpl(final String currentDir,
+                                                     final String jrubyHome,
+                                                     final String[] args) {
+        final RubyInstanceConfig config = new RubyInstanceConfig();
+        if (currentDir != null) config.setCurrentDirectory(currentDir);
+        config.setJRubyHome(jrubyHome);
+        config.processArguments(args);
+        return config;
+    }
+
```
[the remark this answers is not in the document]
Comment (kares) [reply logstash-core/src/main/java/org/logstash/Logstash.java]: yes it can be smt else, in theory e.g. a `classpath://...` like entry.
however should be fine with a `Path` type (for now) on the public methods - good point, updated!

File: logstash-core/src/test/java/org/logstash/AccessorsTest.java (hunk 6 above)
Comment (kares) [reply logstash-core/src/test/java/org/logstash/AccessorsTest.java:35]: > Would it be possible to have this behaviour powered by an annotation instead of subclassing?

yes it would be possible - there's simply an existing convention for some of the IR "ruby-aware" tests 
to use [`RubyEnvTestCase.java`](https://github.com/elastic/logstash/blob/v7.14.0/logstash-core/src/test/java/org/logstash/config/ir/RubyEnvTestCase.java#L33-L39) as a super-class, so here we're following the same ...

also considered the other option of having a JUnit style `@BeforeClass` boot but that still would depend on order or hooks since the `RubyEnvTestCase.java` is using one where it assumes `RubyUtil.RUBY` already booted.

> But I'm not sure that a superclass is the best path. What happens when a new test that uses jruby bits fails to subclass this testing base? Do we end up with wildly-broken behaviour that is easy to trace down, or subtly-broken behaviour that continues to work unless the right edge-case is hit? Is there anything we can do to ensure that we get the former instead of the latter?

Adding/changing an existing Java test (which would be using `org.jruby` bits directly or indirectly and forgot to sub-class or annotate) would lead to an [`AssertionError`](https://github.com/elastic/logstash/pull/13159/files#diff-0e1d049630fdb80df50e27e2bb79fd40c2c0b1ee28c614598e49a72931cfe5b4R59) when initializing the base class (or annotation) -> all classes that depend on the class than lead to a `NoClassDefFoundError` ... here's a [sample CI output where I forgot to sub-class one case](https://logstash-ci.elastic.co/job/elastic+logstash+pull-request+multijob-java-unit-tests/3689/consoleFull)

whether it's an annotation or sub-class does not make much of a difference (in terms of the failure, I think) ... 
if you feel it still provides readability to have an annotation instead I am happy to refactor :wink:

File: logstash-core/src/test/java/org/logstash/config/ir/RubyEnvTestCase.java (hunk 9 above)
Comment (kares) [reply logstash-core/src/test/java/org/logstash/config/ir/RubyEnvTestCase.java:45]: honestly - this whole piece could use a refactoring as it's a bit worrying.
would rather drop it (had issues with this running against jruby-head) - but decided it's out-of-scope here.
there isn't anything new here - the `vendor/bundle/jruby` wasn't validated before ...
what's new is that we no longer have to rely on a system property being set to resolve the path.

File: lib/bootstrap/rspec.rb [hunk 12]
```diff
@@ -20,13 +20,18 @@
 require "logstash-core"
 require "logstash/environment"
 
-$LOAD_PATH.unshift(File.join(LogStash::Environment::LOGSTASH_CORE, "spec"))
+# Bundler + gemspec already setup $LOAD_PATH << '.../lib'
+# but since we load specs from 2 locations we need to hook up these:
```
[the remark this answers is not in the document]
Comment (kares) [reply lib/bootstrap/rspec.rb:24]: kind of yes but the re-use in x-pack is "secondary" in the context here.
for core we load the runtime with CWD == LS_HOME (this is what IDEA does when a test is run from IDE)
for x-pack we leave the CWD as Gradle changes it (to CWD == LS_HOME/x-pack) thus `spec/**/*_spec.rb` will just work.

File: logstash-core/src/test/java/org/logstash/RubyTestBase.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        }
+        LS_HOME = root;
+
+        initializeGlobalRuntime(LS_HOME);
+    }
+
+    static Ruby initializeGlobalRuntime(final Path currentDir) {
+        String[] args = new String[] { "--disable-did_you_mean" };
+        Ruby runtime = Ruby.newInstance(Logstash.initRubyConfig(LS_HOME, currentDir, args));
+        if (runtime != RubyUtil.RUBY) throw new AssertionError("runtime already initialized");
```
Comment (jsvd) [inline logstash-core/src/test/java/org/logstash/RubyTestBase.java:59]:
```suggestion
        if (runtime != RubyUtil.RUBY) throw new AssertionError("runtime already initialized. perhaps a Test class isn't subclassing RubyTestBase as it should?");
```

File: qa/integration/specs/kafka_input_spec.rb [hunk 14]
```diff
@@ -25,27 +25,40 @@
...
-      try(num_retries) do
-        expect(@fixture.output_exists?).to be true
-      end
-
-      try(num_retries) do
-        count = File.foreach(@fixture.actual_output).inject(0) {|c, _| c+1}
-        expect(count).to eq(num_events)
-      end
+    try(num_retries) do
+      expect(File.exists?(file_output_path)).to be(true), "output file: #{file_output_path} does not exist"
```
Comment (jsvd) [inline qa/integration/specs/kafka_input_spec.rb:56]:
```suggestion
      expect(File).to exist(file_output_path) "output file: #{file_output_path} does not exist"
```

File: x-pack/src/test/java/org/logstash/xpack/test/RSpecTests.java [hunk 15]
```diff
@@ -12,24 +12,57 @@
...
+        if (System.getProperty("logstash.root.dir") == null) {
+            // make sure we work from IDE as well as when run with Gradle
+            root = Paths.get("").toAbsolutePath(); // LS_HOME/x-pack
+            if (root.endsWith("x-pack")) root = root.getParent();
+        } else {
+            root = Paths.get(System.getProperty("logstash.root.dir")).toAbsolutePath();
+        }
+
+        if (!root.resolve("versions.yml").toFile().exists()) {
+            throw new AssertionError("versions.yml not found is logstash root dir: " + root);
```
Comment (jsvd) [inline x-pack/src/test/java/org/logstash/xpack/test/RSpecTests.java:41]: Is there an action we'd like the developer to take if this error is encountered?

**Review by jsvd — COMMENTED** [review]
Overall minor comments. I would still like to test this on Windows, when I'm done I'll come back for a final review. Great job here :+1:

**Comment by jsvd** [comment]
also, yay for smaller size:

```
❯ du -sh logstash-*
324M	logstash-8.1.0-SNAPSHOT
297M	logstash-pr13159-8.1.0-SNAPSHOT
165M	logstash-8.1.0-SNAPSHOT.tar.gz
139M	logstash-pr13159-8.1.0-SNAPSHOT.tar.gz
```

File: x-pack/src/test/java/org/logstash/xpack/test/RSpecTests.java (hunk 15 above)
Comment (kares) [reply x-pack/src/test/java/org/logstash/xpack/test/RSpecTests.java:41]: so we previously did not care about what CWD is, the assert is there to fail early. 
it's really an assert the expectation is a Java test is run from an IDE -> cwd is `logstash`, or from gradle in which case `logstash.root.dir` property is properly set. this should be explanatory by following the `AssertionError` all the way up here (and looking at the code above) - :thinking: if we need to explain all of this in the error message here ...

File: logstash-core/build.gradle (hunk 5 above)
Comment (kares) [reply logstash-core/build.gradle:154]: > custom.jruby.path would need more work to get right,

was wrong here - the `custom.jruby.path` bootstrap mechanism in the base `build.gradle` simply skips downloading JRuby and will (build and) copy a `vendor/jruby` from the locally build jruby artifact. there's no special need to select the proper *jruby-complete.jar* from the custom location since the *logstash-core* now relies on the `vendor/jruby/lib/jruby.jar` as an api dependency.

thus actually in case of a `./gradlew clean test -Pcustom.jruby.path=...` everything operates as before

**Comment by kares** [comment]
finally, managed to do a Windows tests (using this branch) - running LS seems to be working as expected:

win-no-jruby-complete

**Comment by kares** [comment]
unfortunately things got :red_circle: *again* due changes upstream - this time the introduction of `JavaVersionCheck`

**Comment by kares** [comment]
jenkins test this again please

**Review by jsvd — APPROVED** [review]
LGTM 🚢 :shipit:
